test/base-type.cpp

```diff
@@ -44,4 +46,33 @@ BOOST_AUTO_TEST_CASE(instantiate)
 	BOOST_CHECK(p);
 }
 
+BOOST_AUTO_TEST_CASE(sort_by_load_after)
+{
+	int totalDependencies{0};
+	std::unordered_set<Type*> previousTypes;
+	for (auto type : Type::GetConfigTypesSortedByLoadDependencies()) {
+		BOOST_CHECK_EQUAL(true, ConfigObject::TypeInstance->IsAssignableFrom(type));
+
+		totalDependencies += type->GetLoadDependencies().size();
+		for (Type* dependency : type->GetLoadDependencies()) {
+			// Note, Type::GetConfigTypesSortedByLoadDependencies() does not detect any cyclic load dependencies,
+			// instead, it relies on this test case to fail.
+			BOOST_CHECK_MESSAGE(previousTypes.find(dependency) != previousTypes.end(), "type '" << type->GetName()
+				<< "' depends on '"<< dependency->GetName() << "' type, but it's not loaded before");
+		}
+
+		previousTypes.emplace(type.get());
+	}
+
+	// The magic number 12 is the sum of host,service,comment,downtime and endpoint load_after dependencies.
+	BOOST_CHECK_MESSAGE(totalDependencies >= 12, "total size of load dependency must be at least 12");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Host::TypeInstance.get()) != previousTypes.end(), "Host type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Service::TypeInstance.get()) != previousTypes.end(), "Service type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Downtime::TypeInstance.get()) != previousTypes.end(), "Downtime type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Comment::TypeInstance.get()) != previousTypes.end(), "Comment type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Notification::TypeInstance.get()) != previousTypes.end(), "Notification type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Zone::TypeInstance.get()) != previousTypes.end(), "Zone type should be in the list");
+	BOOST_CHECK_MESSAGE(previousTypes.find(Endpoint::TypeInstance.get()) != previousTypes.end(), "Endpoint type should be in the list");
```

Comment:
What about just comparing to all types|grep config type? 🙈

Reply:
I am looking for config types that have to be always there, and unlike other types they cannot be disabled with cmake build flags like `-DICINGA2_WITH_ICINGADB=OFF`.

Reply:
> What about just comparing to all types|grep config type? 🙈

Do you mean actually spelling out all types here instead of the currently chosen subset? Or just filtering `Type::GetAllTypes()` and comparing that the same types are contained?

The first approach won't work properly as Yonas already said: the set of registered types is different based on build flags (and actually differs between our Linux and Windows builds).

The latter would work, and actually checking this additionally wouldn't be a bad idea, but some fixed "does it contain this specific type" checks should remain as this serves as a sanity check that some interesting types with dependencies were actually linked into the test binary. If the test binary was built without linking anything expect `lib/base/`, the set of registered types might be too small to contain anything useful as there's actually no `load_after` defined inside `lib/base/*.ti` and the test could actually become pretty pointless. So this is basically the defense against future changes to our CMakeLists accidentally rendering this test useless.